File: generated_code/GPT4/Java/025_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
            left = right = parent = null;
        }
    }

    private Node root;

    public SplayTree() {
        root = null;
    }

    // Public insert method
    public void insert(int key) {
        Node z = root;
        Node p = null;

        while (z != null) {
            p = z;
            if (key < z.key) {
                z = z.left;
            } else if (key > z.key) {
                z = z.right;
            } else {
                // Duplicate keys not allowed, splay existing node
                splay(z);
                return;
            }
        }

        Node n = new Node(key);
        n.parent = p;

        if (p == null) {
            root = n;
        } else if (key < p.key) {
            p.left = n;
        } else {
            p.right = n;
        }

        splay(n);
    }

    // Public delete method
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null) {
            return;
        }

        splay(node);

        if (node.left != null) {
            Node maxLeft = subtreeMaximum(node.left);
            splay(maxLeft);
            maxLeft.right = node.right;
            if (node.right != null)
                node.right.parent = maxLeft;
            root = maxLeft;
            maxLeft.parent = null;
        } else if (node.right != null) {
            root = node.right;
            node.right.parent = null;
        } else {
            root = null;
        }
    }

    // Public search method
    public Integer search(int key) {
        Node node = findNode(key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // Helper: find node with key
    private Node findNode(int key) {
        Node z = root;
        while (z != null) {
            if (key < z.key) {
                z = z.left;
            } else if (key > z.key) {
                z = z.right;
            } else {
                return z;
            }
        }
        return null;
    }

    // Helper: subtree maximum
    private Node subtreeMaximum(Node node) {
        while (node.right != null)
            node = node.right;
        return node;
    }

    // Splay operation: brings node x to root
    private void splay(Node x) {
        if (x == null)
            return;
        while (x.parent != null) {
            if (x.parent.parent == null) {
                // Zig
                if (x.parent.left == x) {
                    rightRotate(x.parent);
                } else {
                    leftRotate(x.parent);
                }
            } else if (x.parent.left == x && x.parent.parent.left == x.parent) {
                // Zig-zig
                rightRotate(x.parent.parent);
                rightRotate(x.parent);
            } else if (x.parent.right == x && x.parent.parent.right == x.parent) {
                // Zig-zig
                leftRotate(x.parent.parent);
                leftRotate(x.parent);
            } else if (x.parent.left == x && x.parent.parent.right == x.parent) {
                // Zig-zag
                rightRotate(x.parent);
                leftRotate(x.parent);
            } else {
                // Zig-zag
                leftRotate(x.parent);
                rightRotate(x.parent);
            }
        }
        root = x;
    }

    // Left rotate at node x
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y == null)
            return;
        x.right = y.left;
        if (y.left != null)
            y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.left = x;
        x.parent = y;
    }

    // Right rotate at node x
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y == null)
            return;
        x.left = y.right;
        if (y.right != null)
            y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.right = x;
        x.parent = y;
    }

    // (Optional) For testing: inorder traversal
    public void inorder() {
        inorderHelper(root);
        System.out.println();
    }

    private void inorderHelper(Node node) {
        if (node != null) {
            inorderHelper(node.left);
            System.out.print(node.key + " ");
            inorderHelper(node.right);
        }
    }
}

File: generated_code/GPT4/Python/025_SplayTree.py
class SplayTree:
    class Node:
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    # Rotate node x with its parent
    def _rotate(self, x):
        p = x.parent
        if not p:
            return
        gp = p.parent
        if p.left == x:
            b = x.right
            x.right = p
            p.left = b
            if b:
                b.parent = p
        else:
            b = x.left
            x.left = p
            p.right = b
            if b:
                b.parent = p
        x.parent = gp
        p.parent = x
        if gp:
            if gp.left == p:
                gp.left = x
            else:
                gp.right = x
        else:
            self.root = x

    # Splay operation: move x to root
    def _splay(self, x):
        while x.parent:
            p = x.parent
            gp = p.parent
            if not gp:
                # Zig
                self._rotate(x)
            elif (gp.left == p and p.left == x) or (gp.right == p and p.right == x):
                # Zig-zig
                self._rotate(p)
                self._rotate(x)
            else:
                # Zig-zag
                self._rotate(x)
                self._rotate(x)

    # Find node with key, or last accessed node (parent if not found)
    def _find(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key < node.key:
                if node.left:
                    node = node.left
                else:
                    break
            elif key > node.key:
                if node.right:
                    node = node.right
                else:
                    break
            else:
                break
        return node, last

    # Public search: splay the found node or last accessed node (parent if not found)
    def search(self, key):
        node, last = self._find(key)
        if node and node.key == key:
            self._splay(node)
            return True
        elif last:
            self._splay(last)
            return False
        else:
            return False

    # Public insert
    def insert(self, key):
        if not self.root:
            self.root = self.Node(key)
            return
        node, last = self._find(key)
        if node and node.key == key:
            self._splay(node)
            return  # Already exists
        new_node = self.Node(key)
        if key < last.key:
            last.left = new_node
            new_node.parent = last
        else:
            last.right = new_node
            new_node.parent = last
        self._splay(new_node)

    # Find maximum node in subtree
    def _subtree_max(self, node):
        while node.right:
            node = node.right
        return node

    # Public delete
    def delete(self, key):
        node, last = self._find(key)
        if not node or node.key != key:
            if last:
                self._splay(last)
            return  # Not found

        self._splay(node)  # Splay the node to root
        left_sub = node.left
        right_sub = node.right

        if left_sub:
            left_sub.parent = None
        if right_sub:
            right_sub.parent = None

        if not left_sub:
            self.root = right_sub
        else:
            max_left = self._subtree_max(left_sub)
            self._splay(max_left)
            max_left.right = right_sub
            if right_sub:
                right_sub.parent = max_left
            self.root = max_left

    # For testing and visualization (not required by prompt)
    def inorder(self):
        result = []
        def _inorder(node):
            if not node:
                return
            _inorder(node.left)
            result.append(node.key)
            _inorder(node.right)
        _inorder(self.root)
        return result